easybuild/tools/filetools.py

```diff
@@ -260,75 +261,34 @@ def download_file(filename, url, path):
     attempt_cnt = 0
     while not downloaded and attempt_cnt < 3:
-        # get HTTP response code first before downloading file
-        response_code = None
-        try:
-            urlfile = urllib.urlopen(url)
-            if hasattr(urlfile, 'getcode'):  # no getcode() in Py2.4 yet
-                response_code = urlfile.getcode()
-            urlfile.close()
-        except IOError, err:
-            _log.warning("Failed to get HTTP response code for %s, retrying: %s", url, err)
-
-        if response_code is not None:
-            _log.debug('HTTP response code for given url: %d', response_code)
-            # check for a 4xx response code which indicates a non-existing URL
-            if response_code // 100 == 4:
-                _log.warning('url %s was not found (HTTP response %d), not trying again', url, response_code)
-                return None
-
-        # use this functions's scope for variables we share with inner function used as report hook for urlretrieve
-        download_file.last_time = time.time()
-        download_file.last_block = 0
-
-        httpmsg = None
-        try:
-            (_, httpmsg) = urllib.urlretrieve(url, path, reporthook=report)
-            _log.info("Downloaded file %s from url %s to %s", filename, url, path)
-
-            if httpmsg.type == "text/html" and not filename.endswith('.html'):
-                _log.warning("HTML file downloaded to %s, so assuming invalid download, retrying.", path)
-                remove_file(path)
-            else:
-                # successful download
+        with open(path, "wb+") as dest_fd:
+            try:
+                src_fd = urllib2.urlopen(url)
+                _log.debug('HTTP response code for given url: %d', src_fd.getcode())
+                dest_fd.write(src_fd.read())
+                _log.info("Downloaded file %s from url %s to %s", filename, url, path)
                 downloaded = True
-        except IOError, err:
-            _log.warning("Error when downloading from %s to %s (%s), removing it and retrying", url, path, err)
-            remove_file(path)
-
-        if not downloaded:
-            attempt_cnt += 1
-            _log.warning("Downloading failed at attempt %s, retrying...", attempt_cnt)
+                src_fd.close()
+            except (ValueError, ) as err:
+                attempt_cnt += 1
+                shutil.copy(url, path)
+                downloaded = True
+            except (urllib2.HTTPError, ) as err:
+                if err.code == 404:
+                    attempt_cnt += 1
+                    _log.warning("Downloading failed at attempt %s, retrying...", attempt_cnt)
+                    continue
+                raise
+            except (IOError, ) as err:
+                if attempt_cnt <= 3:
+                    _log.warning("Failed to get HTTP response code for %s, retrying: %s", url, err)
+                    attempt_cnt += 1
+                    continue
+                raise
```

Comment:
Right, you don't want to raise, you want to print the warning message that the download failed.